integrations/sonarqube/client.py

```diff
@@ -165,56 +156,48 @@ async def send_paginated_api_request(
                 )
-                return all_resources
 
             if e.response.status_code == 404:
                 logger.error(f"Resource not found: {e.response.text}")
-                return all_resources
+
             raise
         except httpx.HTTPError as e:
             logger.error(f"HTTP occurred while fetching paginated data: {e}")
             raise
 
+    @cache_iterator_result()
     async def get_components(
-        self, api_query_params: Optional[dict[str, Any]] = None
-    ) -> list[dict[str, Any]]:
+        self,
+        query_params: Optional[dict[str, Any]] = None,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
         """
         Retrieve all components from SonarQube organization.
 
         :return: A list of components associated with the specified organization.
         """
-        query_params = {}
         if self.organization_id:
-            query_params["organization"] = self.organization_id
             logger.info(
                 f"Fetching all components in organization: {self.organization_id}"
             )
 
-        ## Handle api_query_params based on environment
         if not self.is_onpremise:
             logger.warning(
-                f"Received request to fetch SonarQube components with api_query_params {api_query_params}. Skipping because api_query_params is only supported on on-premise environments"
+                f"Received request to fetch SonarQube components with query_params {query_params}. Skipping because api_query_params is only supported on on-premise environments"
             )
-        else:
-            if api_query_params:
-                query_params.update(api_query_params)
-            elif event.resource_config:
-                # This might be called from places where event.resource_config is not set
-                # like on_start() when creating webhooks
-
-                selector = cast(CustomSelector, event.resource_config.selector)
-                query_params.update(selector.generate_request_params())
 
         try:
-            response = await self.send_paginated_api_request(
-                endpoint=Endpoints.PROJECTS,
+            async for components in self._send_paginated_request(
+                endpoint=Endpoints.COMPONENTS,
                 data_key="components",
+                method="GET",
                 query_params=query_params,
-            )
-            logger.info(
-                f"Fetched {len(response)} components {[item.get("key") for item in response]} from SonarQube"
-            )
-            return response
+            ):
+                logger.info(
+                    f"Fetched {len(components)} components {[item.get('key') for item in components]} from SonarQube"
+                )
+                yield await asyncio.gather(
+                    *[self.get_single_project(project) for project in components]
+                )
```

Comment:
should we be afraid from rate limiting here?

Reply:
No, SonarQube has no rate limit in place. I have included this in the client docs.